### PR TITLE
Obfuscation techniques for android

### LINKED ISSUE #1654: Obfuscation techniques for Android native code
Add recommendations about obfuscation for Android native code?

- Replace libc APIs with syscalls
- Replace libc APIs with custom implementation
- Native code obfuscation using O-LLVM

See: https://darvincitech.wordpress.com/2020/01/07/security-hardening-of-android-native-code/

> All 3 mechanisms are applied on the github project created for Frida detection and the obfuscated APK is provided for interested reverse engineers to analyse. https://github.com/darvincisec/DetectFrida

### PR DESCRIPTION
**Pull request by su-vikas**
Added information about android code hardening techniques to Android Anti-Reversing Defenses chapter. 

If your PR is related to an issue. Please end your PR test with the following line:
This PR closes #1654.

### CODE REVIEW

File: Document/0x05j-Testing-Resiliency-Against-Reverse-Engineering.md [hunk 1]
```diff
@@ -1010,7 +1010,9 @@ Attempt to decompile the bytecode, disassemble any included library files, and p
 
 - meaningful identifiers, such as class names, method names, and variable names, have been discarded,
 - string resources and strings in binaries are encrypted,
-- code and data related to the protected functionality is encrypted, packed, or otherwise concealed.
+- code and data related to the protected functionality is encrypted, packed, or otherwise concealed,
+- for native code, OS syscalls is used, instead of corresponding LibC APIs (e.g open, read),
+- additional obfuscation tool is used to alter the original control flow graph, like [Obfuscator-LLVM](https://github.com/obfuscator-llvm/obfuscator "Obfuscator-LLVM").
 
 For a more detailed assessment, you need a detailed understanding of the relevant threats and the obfuscation methods used.
```
Comment (cpholguera) [inline Document/0x05j-Testing-Resiliency-Against-Reverse-Engineering.md:1017]:
```suggestion
Some of these techniques are discussed and analyzed in the blog post ["Security hardening of Android native code"](https://darvincitech.wordpress.com/2020/01/07/security-hardening-of-android-native-code/) by Gautam Arvind.

For a more detailed assessment, you need a detailed understanding of the relevant threats and the obfuscation methods used. There are some tools such as [APKiD](https://github.com/rednaga/APKiD) that might be able to give you some indications about the kind of obfuscators being used.
```

**Review by cpholguera — CHANGES_REQUESTED** [review]
Hi @su-vikas, thanks a lot for the PR. Please take a look at the suggestions, for some I could not add them as "GitHub Suggestion" unfortunately. Let me know what you think :)

**Comment by su-vikas** [comment]
@cpholguera thanks for the suggestions. I have updated it. I hope i got the separately suggested change right.

File: Document/0x05j-Testing-Resiliency-Against-Reverse-Engineering.md [hunk 2]
```diff
@@ -1006,13 +1006,18 @@ In the test case "Make Sure That Free Security Features Are Activated (MSTG-CODE
...
-- code and data related to the protected functionality is encrypted, packed, or otherwise concealed.
+- Obfuscation often carries a cost in runtime performance, therefore it might have been only applied to certain very specific parts of the code, typically those dealing with security and runtime protection.
+- Meaningful identifiers, such as class names, method names, and variable names, might have been discarded.
+- String resources and strings in binaries might have been encrypted.
+- Code and data related to the protected functionality might be encrypted, packed, or otherwise concealed.
+- for native code, [libc APIs](https://man7.org/linux/man-pages/dir_section_3.html) (e.g open, read) might have been replaced with OS [syscalls](https://man7.org/linux/man-pages/man2/syscalls.2.html).
+- additional obfuscation techniques such as ["Control Flow Flattening"](https://github.com/obfuscator-llvm/obfuscator/wiki/Control-Flow-Flattening) or ["Bogus Control Flow"](https://github.com/obfuscator-llvm/obfuscator/wiki/Bogus-Control-Flow) might have been applied using e.g. [Obfuscator-LLVM](https://github.com/obfuscator-llvm/obfuscator "Obfuscator-LLVM").
```
Comment (cpholguera) [inline Document/0x05j-Testing-Resiliency-Against-Reverse-Engineering.md:1016]:
```suggestion
- Additional obfuscation techniques such as ["Control Flow Flattening"](https://github.com/obfuscator-llvm/obfuscator/wiki/Control-Flow-Flattening) or ["Bogus Control Flow"](https://github.com/obfuscator-llvm/obfuscator/wiki/Bogus-Control-Flow) might have been applied using e.g. [Obfuscator-LLVM](https://github.com/obfuscator-llvm/obfuscator "Obfuscator-LLVM").
```

File: Document/0x05j-Testing-Resiliency-Against-Reverse-Engineering.md [hunk 3]
```diff
@@ -1006,13 +1006,18 @@ In the test case "Make Sure That Free Security Features Are Activated (MSTG-CODE
...
+Attempt to decompile the bytecode, disassemble any included library files and try to understand it. When doing so, consider the following:
 
-- meaningful identifiers, such as class names, method names, and variable names, have been discarded,
-- string resources and strings in binaries are encrypted,
-- code and data related to the protected functionality is encrypted, packed, or otherwise concealed.
+- Obfuscation often carries a cost in runtime performance, therefore it might have been only applied to certain very specific parts of the code, typically those dealing with security and runtime protection.
+- Meaningful identifiers, such as class names, method names, and variable names, might have been discarded.
+- String resources and strings in binaries might have been encrypted.
+- Code and data related to the protected functionality might be encrypted, packed, or otherwise concealed.
+- for native code, [libc APIs](https://man7.org/linux/man-pages/dir_section_3.html) (e.g open, read) might have been replaced with OS [syscalls](https://man7.org/linux/man-pages/man2/syscalls.2.html).
```
Comment (cpholguera) [inline Document/0x05j-Testing-Resiliency-Against-Reverse-Engineering.md:1015]:
```suggestion
- For native code, [libc APIs](https://man7.org/linux/man-pages/dir_section_3.html) (e.g open, read) might have been replaced with OS [syscalls](https://man7.org/linux/man-pages/man2/syscalls.2.html).
```

**Review by cpholguera — APPROVED** [review]
Awesome, thanks a lot @su-vikas!

**Pull request merged**